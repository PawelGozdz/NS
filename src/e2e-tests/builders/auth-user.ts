import { Kysely, Transaction } from 'kysely';

import { TableNames, TestingE2EFunctions, dialect, kyselyPlugins } from '@app/core';
import { IDatabaseModels } from '@libs/common';
import { AuthenticationServer } from '@libs/testing';

import { UserSeedBuilder } from './user-builder';

const tablesInvolved = [TableNames.USERS, TableNames.AUTH_USERS, TableNames.USER_PROFILES];

const authenticationServer = new AuthenticationServer();

type IDatabaseDaos = IDatabaseModels;

const dbConnection = new Kysely<IDatabaseDaos>({
  dialect,
  plugins: kyselyPlugins,
});
const dbUtils = new TestingE2EFunctions(dbConnection);

export const getAccessToken = () => authenticationServer.generateAccessToken();
export const getRefreshToken = () => authenticationServer.generateRefreshToken();

export const getCookies = () =>
  authenticationServer.getTokensAsCookie({
    accessToken: getAccessToken(),
    refreshToken: getRefreshToken(),
  });

export async function loginUser(transaction?: Transaction<IDatabaseDaos>) {
  if (transaction) {
    return seed(transaction);
  }
  return dbConnection.transaction().execute(async (trx) => seed(trx));
}

async function seed(trx: Transaction<IDatabaseDaos>) {
  await dbUtils.truncateTables(tablesInvolved, trx);

  const seedBuilder = await UserSeedBuilder.create(trx);
  seedBuilder.withUser().withAuthUser().withProfile();
  await seedBuilder.build();
  return seedBuilder;
}
