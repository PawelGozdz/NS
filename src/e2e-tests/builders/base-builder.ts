/* eslint-disable @typescript-eslint/strict-boolean-expressions */
/* eslint-disable @typescript-eslint/prefer-readonly */
/* eslint-disable @typescript-eslint/no-explicit-any */
// src/seed-builders/base-seed-builder.ts
import { Kysely, SelectExpression, sql, Transaction } from 'kysely';
import { InsertExpression } from 'kysely/dist/cjs/parser/insert-values-parser';
import { chunk, sampleSize, take } from 'lodash';

import { IDatabaseModels, TableNames } from '@app/core';

import { GetOptions } from './interfaces';

type GetDaosOptions = {
  limit?: number;
  random?: boolean;
} & ({ first?: true; last?: false } | { first?: false; last?: true } | { first?: false; last?: false });

export abstract class BaseSeedBuilder<T> {
  protected dbConnection: Kysely<IDatabaseModels> | Transaction<IDatabaseModels>;

  protected tableName: TableNames;

  protected daos: T[] = [];

  constructor(dbConnection: Kysely<IDatabaseModels> | Transaction<IDatabaseModels>, tableName: TableNames) {
    this.dbConnection = dbConnection;
    this.tableName = tableName;
  }

  getDaos(options?: GetDaosOptions): T[] {
    let result = this.daos;

    if (options?.random) {
      result = sampleSize(result, options.limit ?? result.length);
    } else if (options?.first) {
      result = take(result, options.limit ?? 1);
    } else if (options?.last) {
      result = take(result.reverse(), options.limit ?? 1);
    } else if (options?.limit) {
      result = take(result, options.limit);
    }

    return result;
  }

  public getTableName() {
    return this.tableName;
  }

  async getRandomIndex(options?: GetOptions): Promise<number | string> {
    const indexes = await this.selectFields({
      ...options,
      limit: 1,
      shuffle: true,
    });
    if (indexes.length === 0) {
      throw new Error(`No records found in table ${this.getTableName()}`);
    }
    return indexes[0].id;
  }

  async getFirst(options?: Pick<GetOptions, 'fields' | 'sort'>): Promise<T> {
    const sort = {
      sortColumn: options?.sort?.sortColumn ?? 'id',
      sortDirection: 'asc' as const,
    };

    const indexes = await this.selectFields({
      limit: 1,
      shuffle: false,
      sort,
    });
    if (indexes.length === 0) {
      throw new Error(`No records found in table ${this.getTableName()}`);
    }
    return indexes[0];
  }

  async getLast(options?: Pick<GetOptions, 'fields' | 'sort'>): Promise<T> {
    const sort = {
      sortColumn: options?.sort?.sortColumn ?? 'id',
      sortDirection: 'desc' as const,
    };

    const indexes = await this.selectFields({
      limit: 1,
      sort,
    });
    if (indexes.length === 0) {
      throw new Error(`No records found in table ${this.getTableName()}`);
    }

    return indexes[0];
  }

  async getMany(qty: number, options?: GetOptions): Promise<T[]> {
    const indexes = await this.selectFields({ ...options, limit: qty });

    if (indexes.length === 0) {
      throw new Error(`No records found in table ${this.getTableName()}`);
    }

    let result = indexes;

    if (options?.stretchRandom) {
      result = Array.from({ length: qty }, () => result[Math.floor(Math.random() * result.length)]);
    } else if (options?.stretch && result.length < qty) {
      result = Array.from({ length: qty }, (_, i) => result[i % result.length]);
    }

    return result;
  }

  async selectFields(options: GetOptions): Promise<any[]> {
    const fields = options?.fields ? options.fields : ['id'];

    let query = this.dbConnection.selectFrom(this.getTableName()).select(fields as unknown as SelectExpression<IDatabaseModels, TableNames>);

    if (options?.shuffle) {
      query = query.orderBy(sql`RANDOM()`);
    }

    if (options?.sort) {
      query = query.orderBy(options.sort.sortColumn, options.sort.sortDirection);
    }

    if (options?.limit) {
      query = query.limit(options.limit);
    }
    return query.execute();
  }

  async insertDaos(values: T[]) {
    const models: T[] = [];

    for await (const chunks of chunk(values, 10000)) {
      models.push(...(await this.insertSmallChunk(chunks as InsertExpression<IDatabaseModels, TableNames>)));
    }

    return models;
  }

  private async insertSmallChunk(values: InsertExpression<IDatabaseModels, TableNames>) {
    return (await this.dbConnection.insertInto(this.getTableName()).values(values).returningAll().execute()) as T[];
  }
}
