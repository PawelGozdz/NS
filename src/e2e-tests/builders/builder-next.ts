import { Database } from '@app/database';
import { IDatabaseModels } from '@libs/common';
import { Transaction, sql } from 'kysely';

interface RelationConfig {
	foreignKey: string;
	primaryKey: string;
	parentEntity: string;
}

interface SeederConfig {
	table: string;
	data: (parent: any, index: number) => Record<string, any>;
	quantity?: number;
	return?: string[] | ((parent: any) => string[]);
	order?: number;
	relations?: { config: SeederConfig; relation: RelationConfig }[];
}

export class DataSeeder {
	constructor(private db: Database) {}

	async seed(configs: SeederConfig[]): Promise<void> {
		configs.sort((a, b) => (a.order || 0) - (b.order || 0));
		for (const config of configs) {
			await this.db.transaction().execute(async (transaction) => {
				await this.seedRecursive(config, transaction);
			});
		}
	}

	private async seedRecursive(config: SeederConfig, transaction: Transaction<IDatabaseModels>, parent?: any, index?: number) {
		console.log('PAREMT', await parent, config);
		const data = Array(config.quantity || 1)
			.fill(0)
			.map((_, i) => (config.data as (parent: any, index: number) => Record<string, any>)(parent, i));

		const returnFields = typeof config.return === 'function' ? config.return(parent) : config.return;

		if (parent) {
			// w childernie data jes toromisem i nie filtruje indeined
			console.log('DATA', data);
		}

		const filteredData = await Promise.all(this.filterUndefinedProperties(data));
		console.log('FILTERED DATA', filteredData);

		const inserted = await transaction
			.insertInto(config.table as keyof IDatabaseModels)
			.values(filteredData as any)
			.returning(returnFields as any)
			.execute();

		console.log('INSERT', inserted);
		console.log('CONFIG RLEATIOSN____________________________', config.relations);
		if (config.relations) {
			for (const { config: relationConfig, relation } of config.relations) {
				if (relation.parentEntity !== config.table) {
					throw new Error(`Invalid relation: ${relation.parentEntity} is not the parent of ${relationConfig.table}`);
				}
				if ((relationConfig.order || 0) <= (config.order || 0)) {
					throw new Error(
						`Invalid order: The order of ${relationConfig.table} cannot be less than or equal to the order of its parent ${config.table}`,
					);
				}
				for (let i = 0; i < inserted.length; i++) {
					// { id: 'a6185a9f-8873-4f1b-b630-3729318bc636', email: 'abc@co' }
					const buildParent = inserted[i];
					// AuthUser dao
					let relationData = relationConfig.data(buildParent, i);
					// console.log('RELATIOAN DATA', relationData);

					relationData[relation.foreignKey] = buildParent[relation.primaryKey];
					// console.log('REL DATA', relationData);
					// relationConfig.data = relationData;
					if (parent) {
						console.log(relationConfig, buildParent, i);
					}
					await this.seedRecursive(relationConfig, transaction, buildParent, i);
				}
			}
		}
		// console.log('INESRE', inserted);

		return inserted;
	}

	async rollback(configs: SeederConfig[]): Promise<void> {
		configs.sort((a, b) => (b.order || 0) - (a.order || 0));
		for (const config of configs) {
			await this.db.transaction().execute(async (transaction) => {
				await this.rollbackRecursive(config, transaction);
			});
		}
	}

	private async rollbackRecursive(config: SeederConfig, transaction: Transaction<IDatabaseModels>): Promise<void> {
		if (config.relations) {
			for (const { config: relationConfig } of config.relations) {
				await this.rollbackRecursive(relationConfig, transaction);
			}
		}

		await sql`DELETE FRIM ${config.table}`.execute(transaction);
	}

	private filterUndefinedProperties(array: Record<string, any>[]) {
		return array.map((obj) => {
			const filteredObj = {};
			Object.entries(obj).forEach(([key, value]) => {
				if (value !== undefined) {
					filteredObj[key] = value;
				}
			});
			return filteredObj;
		});
	}
}
