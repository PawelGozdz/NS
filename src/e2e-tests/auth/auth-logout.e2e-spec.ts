import { INestApplication } from '@nestjs/common';
import { Test, TestingModule } from '@nestjs/testing';
import { Kysely } from 'kysely';
import request from 'supertest';

import { HashService } from '@app/contexts';
import { ApiResponseStatusJsendEnum, AppRoutes, IDatabaseModels, dialect, kyselyPlugins } from '@app/core';
import { UnauthorizedError } from '@libs/common';

import { AppModule } from '../../app.module';
import { getCookies, loginUser } from '../builders/auth-user';

type IDdbDaos = IDatabaseModels;

describe('AuthJwtControllerV1 -> logout (e2e)', () => {
  const dbConnection = new Kysely<IDdbDaos>({
    dialect,
    plugins: kyselyPlugins,
  });
  let app: INestApplication;

  beforeAll(async () => {
    const moduleFixture: TestingModule = await Test.createTestingModule({
      imports: [AppModule],
      providers: [HashService],
    }).compile();

    app = moduleFixture.createNestApplication();
    await app.init();
  });

  afterAll(async () => {
    await dbConnection.destroy();
    await app.close();
  });

  let cookies: [string, string];

  beforeEach(async () => {
    cookies = getCookies();

    await dbConnection.transaction().execute(async (trx) => {
      await loginUser(trx);
    });
  });

  describe('/auth/logout (POST) V1', () => {
    describe('SUCCESS', () => {
      it('should return both tokens if refresh token is still valid', async () => {
        // Act
        const response = await request(app.getHttpServer())
          .post(AppRoutes.AUTH.v1.logout)
          .set(...cookies)
          .set('Content-Type', 'application/json')
          .send();

        // Assert
        expect(response.statusCode).toBe(204);
        expect(response.body).toMatchSnapshot();
      });
    });

    describe('FAILURE', () => {
      it('should return UnauthorizedError if incorrect refreshToken', async () => {
        // Act
        const response = await request(app.getHttpServer())
          .post(AppRoutes.AUTH.v1.logout)
          .set('Cookie', 'incorrect-cookie')
          .set('Content-Type', 'application/json')
          .send();

        // Assert
        expect(response.statusCode).toBe(401);
        expect(response.body.status).toBe(ApiResponseStatusJsendEnum.FAIL);
        expect(response.body).toMatchSnapshot({
          timestamp: expect.any(String),
          data: {
            error: UnauthorizedError.message,
          },
        });
      });
    });
  });
});
