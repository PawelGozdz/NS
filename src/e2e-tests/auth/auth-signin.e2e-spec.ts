import { INestApplication } from '@nestjs/common';
import { Test, TestingModule } from '@nestjs/testing';
import { Kysely } from 'kysely';
import request from 'supertest';

import { AppRoutes, TableNames, TestingE2EFunctions, dialect, kyselyPlugins } from '@app/core';
import { ApiResponseStatusJsendEnum, UnauthorizedError } from '@libs/common';
import { TestLoggerModule, testingDefaults } from '@libs/testing';

import { AppModule } from '../../app.module';
import { UserSeedBuilder } from '../builders/user-builder';

type IDdbDaos = { [key: string]: unknown };

describe('AuthJwtControllerV1 -> signin (e2e)', () => {
  const dbConnection = new Kysely<IDdbDaos>({
    dialect,
    plugins: kyselyPlugins,
  });
  const dbUtils = new TestingE2EFunctions(dbConnection);
  let app: INestApplication;

  const tablesInvolved = [TableNames.USERS, TableNames.AUTH_USERS, TableNames.USER_PROFILES];

  beforeAll(async () => {
    const moduleFixture: TestingModule = await Test.createTestingModule({
      imports: [AppModule, TestLoggerModule.forRoot()],
    }).compile();

    app = moduleFixture.createNestApplication();
    await app.init();
  });

  afterAll(async () => {
    await dbConnection.destroy();
    await app.close();
  });

  beforeEach(async () => {
    await dbConnection.transaction().execute(async (trx) => {
      await dbUtils.truncateTables(tablesInvolved, trx);

      const seedBuilder = await UserSeedBuilder.create(trx);
      seedBuilder.withUser().withAuthUser();
      await seedBuilder.build();
    });
  });

  const defaultCorrectEmail = testingDefaults.email;
  const defaultcorrectPassword = testingDefaults.userPassword;

  describe('/auth/signin (POST) V1', () => {
    describe('SUCCESS', () => {
      it('should return success and both tokens', async () => {
        // Act
        const response = await request(app.getHttpServer()).post(AppRoutes.AUTH.v1.signin).set('Content-Type', 'application/json').send({
          email: defaultCorrectEmail,
          password: defaultcorrectPassword,
        });

        // Assert
        expect(response.statusCode).toBe(200);
        expect(response.body.status).toBe(ApiResponseStatusJsendEnum.SUCCESS);

        expect(response.body).toMatchSnapshot({
          timestamp: expect.any(String),
          data: {
            access_token: expect.any(String),
            refresh_token: expect.any(String),
          },
        });
      });
    });

    describe('FAILURE', () => {
      it('should return input validation errors', async () => {
        // Act
        const response = await request(app.getHttpServer()).post(AppRoutes.AUTH.v1.signin).set('Content-Type', 'application/json').send({
          email: null,
          password: 12,
        });

        // Assert
        expect(response.statusCode).toBe(400);
        expect(response.body.status).toBe(ApiResponseStatusJsendEnum.FAIL);
        expect(response.body).toMatchSnapshot({
          timestamp: expect.any(String),
          data: {
            subErrors: expect.any(Array),
            error: expect.any(String),
          },
        });
      });

      it('should return UnauthorizedError', async () => {
        // Act
        const response = await request(app.getHttpServer()).post(AppRoutes.AUTH.v1.signin).set('Content-Type', 'application/json').send({
          email: defaultCorrectEmail,
          password: 'incorrect-password',
        });

        // Assert
        expect(response.statusCode).toBe(401);
        expect(response.body.status).toBe(ApiResponseStatusJsendEnum.FAIL);
        expect(response.body.data?.error).toEqual(UnauthorizedError.message);
        expect(response.body).toMatchSnapshot({
          timestamp: expect.any(String),
        });
      });
    });
  });
});
