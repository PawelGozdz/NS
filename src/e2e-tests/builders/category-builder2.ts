/* eslint-disable no-plusplus */
import { Kysely, Transaction } from 'kysely';

import { CategoryModel } from '@app/contexts';
import { IDatabaseModels, TableNames } from '@app/core';

import { CategoryFixtureFactory } from '../fixtures';
import { BaseSeedBuilder } from './base-builder';
import { InsertOptions } from './interfaces';

export class CategorySeedBuilder2 extends BaseSeedBuilder<CategoryModel> {
  constructor(dbConnection: Kysely<IDatabaseModels> | Transaction<IDatabaseModels>) {
    super(dbConnection, TableNames.CATEGORIES);
  }

  async insert(options: InsertOptions): Promise<void> {
    const quantity = options.quantity ?? 1;

    const values: CategoryModel[] = [];
    for (let i = 0; i < quantity; i++) {
      values.push(CategoryFixtureFactory.createRandom());
    }

    this.daos = await this.insertDaos(values);
  }
}
