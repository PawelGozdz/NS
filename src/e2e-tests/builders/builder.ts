import { AuthUserDao, UserDao } from '@app/contexts/auth';
import { Kysely } from 'kysely';
import { TableNames, dialect, kyselyPlugins } from '../../database';
import { AuthUserFixtureFactory, UserFixtureFactory } from '../fixtures';

type IDatabaseDaos = any;

export class UserSeedBuilder {
	public dbConnection: Kysely<IDatabaseDaos>;
	public userDao: UserDao;
	public authUserDao: AuthUserDao;

	daos: {
		userDaoObj: UserDao | undefined;
		authUserDaoObj: AuthUserDao | undefined;
	} = {
		userDaoObj: undefined,
		authUserDaoObj: undefined,
	};

	actions: { method: string }[] = [];
	cookies: string[] = [];

	private constructor(dbConnection: Kysely<IDatabaseDaos>) {
		this.dbConnection = dbConnection;
	}

	async build() {
		for await (const action of this.actions) {
			await this[action.method]();
		}

		return this;
	}

	static async create(db?: Kysely<IDatabaseDaos>): Promise<UserSeedBuilder> {
		const builder = new UserSeedBuilder(
			db ||
				new Kysely<IDatabaseDaos>({
					dialect,
					plugins: kyselyPlugins,
				}),
		);

		return builder;
	}

	private async insertUser() {
		if (!this.daos.userDaoObj) {
			throw new Error('UserDao is not defined');
		}

		await this.dbConnection.insertInto(TableNames.USERS).values(this.daos.userDaoObj).execute();
		this.userDao = (await this.dbConnection
			.selectFrom(TableNames.USERS)
			.selectAll()
			.where('id', '=', this.daos.userDaoObj.id)
			.executeTakeFirst()) as UserDao;
	}

	withUser(user?: Partial<UserDao>): this {
		this.daos.userDaoObj = UserFixtureFactory.create({
			...user,
		});

		this.actions.push({ method: 'insertUser' });

		return this;
	}

	private async insertAuthUser(): Promise<this> {
		if (!this.daos.authUserDaoObj) {
			throw new Error('AuthUserDao is not defined');
		}

		await this.dbConnection.insertInto(TableNames.AUTH_USERS).values(this.daos.authUserDaoObj).execute();
		this.authUserDao = (await this.dbConnection
			.selectFrom(TableNames.AUTH_USERS)
			.selectAll()
			.where('userId', '=', this.daos.authUserDaoObj.id)
			.executeTakeFirst()) as AuthUserDao;

		return this;
	}

	withAuthUser(authUser?: Partial<AuthUserDao>): this {
		if (!this.daos.userDaoObj) {
			throw new Error('UserDao is not defined');
		}

		this.daos.authUserDaoObj = AuthUserFixtureFactory.create({
			userId: this.daos.userDaoObj.id,
			...authUser,
		});

		this.actions.push({ method: 'insertAuthUser' });

		return this;
	}
}
