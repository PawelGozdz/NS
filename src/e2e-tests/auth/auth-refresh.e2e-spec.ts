import { TestingE2EFunctions } from '@app/database/kysley';
import { ApiResponseStatusJsendEnum } from '@libs/common/api';
import { AuthenticationServer, TestLoggerModule } from '@libs/testing';
import { INestApplication } from '@nestjs/common';
import { Test, TestingModule } from '@nestjs/testing';
import { Kysely } from 'kysely';
import request from 'supertest';

import { HashService } from '@app/contexts/auth';
import { AppRoutes } from '@app/core';
import { UnauthorizedError } from '@libs/common';
import { AppModule } from '../../app.module';
import { TableNames, dialect, kyselyPlugins } from '../../database';
import { UserSeedBuilder } from '../builders/builder';

type IDdbDaos = any;

describe('AuthJwtControllerV1 -> refresh (e2e)', () => {
	const dbConnection = new Kysely<IDdbDaos>({
		dialect,
		plugins: kyselyPlugins,
	});
	const dbUtils = new TestingE2EFunctions(dbConnection);
	let app: INestApplication;
	let authenticationServer: AuthenticationServer;
	let hashService: HashService;

	const tablesInvolved = [TableNames.USERS, TableNames.AUTH_USERS, TableNames.USER_PROFILES];

	beforeAll(async () => {
		const moduleFixture: TestingModule = await Test.createTestingModule({
			imports: [AppModule, TestLoggerModule.forRoot()],
			providers: [HashService],
		}).compile();

		app = moduleFixture.createNestApplication();
		await app.init();

		authenticationServer = new AuthenticationServer();
		hashService = app.get(HashService);
	});

	afterAll(async () => {
		await dbConnection.destroy();
		await app.close();
	});

	let cookieTokens: [string, string];
	let cookies: [string, string];

	beforeEach(async () => {
		cookieTokens = [authenticationServer.generateAccessToken(), authenticationServer.generateRefreshToken()];
		cookies = authenticationServer.getTokensAsCookie({
			accessToken: cookieTokens[0],
			refreshToken: cookieTokens[1],
		});

		await dbConnection.transaction().execute(async (trx) => {
			await dbUtils.truncateTables(tablesInvolved, trx);

			const seedBuilder = await UserSeedBuilder.create(trx);
			seedBuilder.withUser().withAuthUser({
				hashedRt: await hashService.hashData(cookieTokens[1]),
			});
			await seedBuilder.build();
		});
	});

	describe('/auth/refresh (POST) V1', () => {
		describe('SUCCESS', () => {
			it('should return both tokens if refresh token is still valid', async () => {
				// Act
				const response = await request(app.getHttpServer())
					.post(AppRoutes.AUTH.v1.refresh)
					.set(...cookies)
					.set('Content-Type', 'application/json')
					.send();

				// Assert
				expect(response.statusCode).toBe(200);
				expect(response.body.status).toBe(ApiResponseStatusJsendEnum.SUCCESS);
				expect(response.body).toMatchSnapshot({
					timestamp: expect.any(String),
					data: {
						access_token: expect.any(String),
						refresh_token: expect.any(String),
					},
				});
			});
		});

		describe('FAILURE', () => {
			it('should return UnauthorizedError if incorrect refreshToken', async () => {
				// Act
				const response = await request(app.getHttpServer())
					.post(AppRoutes.AUTH.v1.refresh)
					.set('Cookie', 'incorrect-cookie')
					.set('Content-Type', 'application/json')
					.send();

				// Assert
				expect(response.statusCode).toBe(401);
				expect(response.body.status).toBe(ApiResponseStatusJsendEnum.FAIL);
				expect(response.body).toMatchSnapshot({
					timestamp: expect.any(String),
					data: {
						error: UnauthorizedError.message,
					},
				});
			});
		});
	});
});
