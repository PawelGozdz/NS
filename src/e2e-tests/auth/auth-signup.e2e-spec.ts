import { TestingE2EFunctions } from '@app/database/kysley';
import { ApiResponseStatusJsendEnum } from '@libs/common/api';
import { AuthenticationServer, TestLoggerModule, testingDefaults } from '@libs/testing';
import { INestApplication } from '@nestjs/common';
import { Test, TestingModule } from '@nestjs/testing';
import { Kysely } from 'kysely';
import request from 'supertest';

import { UserAlreadyExistsError } from '@app/contexts/auth/users/domain/users/errors/user-already-exists.error';
import { AppRoutes } from '@app/core';
import { AppModule } from '../../app.module';
import { TableNames, dialect, kyselyPlugins } from '../../database';
import { UserSeedBuilder } from '../builders/builder';

type IDdbDaos = any;

describe('AuthJwtControllerV1 -> signup (e2e)', () => {
	const dbConnection = new Kysely<IDdbDaos>({
		dialect,
		plugins: kyselyPlugins,
	});
	const dbUtils = new TestingE2EFunctions(dbConnection);
	let app: INestApplication;
	let authenticationServer: AuthenticationServer;

	const tablesInvolved = [TableNames.USERS, TableNames.USERS];

	beforeAll(async () => {
		const moduleFixture: TestingModule = await Test.createTestingModule({
			imports: [AppModule, TestLoggerModule.forRoot()],
		}).compile();

		app = moduleFixture.createNestApplication();
		await app.init();

		authenticationServer = new AuthenticationServer();
	});

	afterAll(async () => {
		await dbConnection.destroy();
		await app.close();
	});

	beforeEach(async () => {
		await dbConnection.transaction().execute(async (trx) => {
			await dbUtils.truncateTables(tablesInvolved, trx);
		});
	});

	describe('/auth/signup (POST) V1', () => {
		describe('SUCCESS', () => {
			it('should return success and both tokens', async () => {
				// Act
				const response = await request(app.getHttpServer()).post(AppRoutes.AUTH.v1.signup).set('Content-Type', 'application/json').send({
					email: testingDefaults.email,
					password: testingDefaults.userPassword,
				});

				// Assert
				expect(response.statusCode).toBe(201);
				expect(response.body.status).toBe(ApiResponseStatusJsendEnum.SUCCESS);
				expect(response.body).toMatchSnapshot({
					timestamp: expect.any(String),
					data: {
						access_token: expect.any(String),
						refresh_token: expect.any(String),
					},
				});
			});
		});

		describe('FAILURE', () => {
			it('should return input validation errors', async () => {
				// Act
				const response = await request(app.getHttpServer()).post(AppRoutes.AUTH.v1.signup).set('Content-Type', 'application/json').send({
					email: null,
					password: 12,
				});

				// Assert
				expect(response.statusCode).toBe(400);
				expect(response.body.status).toBe(ApiResponseStatusJsendEnum.FAIL);
				expect(response.body).toMatchSnapshot({
					timestamp: expect.any(String),
					data: {
						subErrors: expect.any(Array),
						error: expect.any(String),
					},
				});
			});

			it('should return conflict error', async () => {
				// Arrange
				await dbConnection.transaction().execute(async (trx) => {
					await dbUtils.truncateTables(tablesInvolved, trx);

					const seedBuilder = await UserSeedBuilder.create(trx);
					seedBuilder.withUser().withAuthUser();
					await seedBuilder.build();
				});

				// Act
				const response = await request(app.getHttpServer()).post(AppRoutes.AUTH.v1.signup).set('Content-Type', 'application/json').send({
					email: testingDefaults.email,
					password: testingDefaults.userPassword,
				});

				// Assert
				expect(response.statusCode).toBe(409);
				expect(response.body.status).toBe(ApiResponseStatusJsendEnum.FAIL);
				expect(response.body).toMatchSnapshot({
					timestamp: expect.any(String),
					data: {
						error: UserAlreadyExistsError.withEmail(testingDefaults.email).message,
					},
				});
			});
		});
	});
});
