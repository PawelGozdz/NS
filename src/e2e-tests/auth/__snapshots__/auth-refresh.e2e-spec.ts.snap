// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AuthJwtControllerV1 -> refresh (e2e) /auth/refresh (POST) V1 FAILURE should return UnauthorizedError if incorrect refreshToken 1`] = `
{
  "data": {
    "error": "Unauthorized",
  },
  "path": "/auth/refresh",
  "status": "fail",
  "statusCode": 401,
  "timestamp": Any<String>,
}
`;

exports[`AuthJwtControllerV1 -> refresh (e2e) /auth/refresh (POST) V1 SUCCESS should return both tokens if refresh token is still valid 1`] = `
{
  "data": {
    "access_token": Any<String>,
    "refresh_token": Any<String>,
  },
  "path": "/auth/refresh",
  "status": "success",
  "statusCode": 200,
  "timestamp": Any<String>,
}
`;
